javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+A rock paper scissors game might involve a couple basic things:
+
+- Players' scores
+- The ability to play a round (and playing a round should update a player's score)
+
+And might also include a couple nice-to-haves
+
+- The ability to determine the current winning player
+- The ability to restart the game
+
+So, at it's most basic, an object that represents the game might look something like this (assuming we're playing against a computer player):
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    // code to play the round... (and update the scores when a player wins)
+  }
+}
+```
+
+And if we fleshed it out, our object may come out to look something like this:
+
+```javascript
+const rps = {
+  playerScore: 0,
+  computerScore: 0,
+  playRound(playerChoice) {
+    const options = ['rock', 'paper', 'scissors']
+
+    // if an invalid choice is chosen, throw an error
+    if(!options.includes(playerChoice.toLowerCase())) {
+      throw new Error(`Expected 'rock', 'paper', or 'scissors', but got ${playerChoice}`)
+    }
+    
+    // get the computer's choice
+    const computerChoice = options[Math.floor(Math.random() * 3)]
+
+
+    // determine the winner, apply points if necessary, and return who won
+    if(playerChoice.toLowerCase() === computerChoice) {
+      return "tie"
+    } else if(
+      (playerChoice === 'rock' && computerChoice === 'scissors') ||
+      (playerChoice === 'paper' && computerChoice === 'rock') ||
+      (playerChoice === 'scissors' && computerChoice === 'paper')
+    ) {
+      this.playerScore++
+      return "player"
+    } else {
+      this.computerScore++
+      return 'computer'
+    }
+  },
+  getWinningPlayer() {
+    if(this.playerScore === this.computerScore) {
+      return 'tie'
+    } else if (this.playerScore > this.computerScore) {
+      return 'player'
+    } else {
+      return 'computer'
+    }
+  },
+  reset() {
+    this.playerScore = 0;
+    this.computerScore = 0;
+  }
+}
+
+rps.playRound('rock') // returns 'player' if we win...
+rps.playerScore       // ...and our score would have increased
+
+// We also have the ability to check the winner and reset the game at any time
+rps.getWinningPlayer() // 'player', if we won above round
+rps.reset()
+```
+
+You may be looking at this code and thinking that you personally prefer to split your code between more functions than you see here, but also recognize that those functions may not really be a useful interaction point for anyone using your object.
+
+But, there is no rule saying that you can't add those functions to your object as well! A common convention is to prefix methods and properties that you don't intend other people to use with an underscore (`_`). This convention conveys to others that "These things are meant to be used internally by this object, please interact with the other available methods and properties on this object's interface instead".
```

Comment:
Suggestion: Instead of using underscore to prefix private properties (which is not enforced at runtime at all), imo it'd be better to teach learners about private properties using `#`property syntax

thoughts?

Reply:
☝️ 

It's also not an uncommon thing to see in the wild. I would bet material TOP points to even has a few examples of the pattern being used already. IMO this is a decent way to discuss the basic concept without introducing anything overtly fancy- which is kinda the goal of using object literals in the first place. The concepts here can carry over from here into every other lesson on object creation. 

A lesson note could be added to _stress_ the fact it's just a common convention and is not enforced. I wouldn't be opposed to it, but I personally think that's already pretty clear.